### PR TITLE
Feat/sign for staking

### PR DESCRIPTION
**Pull request by rafaelugolini**
This PR depends on https://github.com/mstable/mStable-api/pull/14

### CODE REVIEW

File: apps/governance/src/app/pages/Stake/StakeForm.tsx [hunk 1]
```diff
@@ -138,6 +140,22 @@ export const StakeForm: FC<Props> = ({ className, isMigrating = false }) => {
...
+
+      const body = await signature.json()
+      setStakeSignatures(prevSignatures => ({
+        ...prevSignatures,
+        [address]: body.signature,
+      }))
+    }
+    fetchSignature()
+  }, [signer, setStakeSignatures])
+
```
Comment (chrisjgf) [inline apps/governance/src/app/pages/Stake/StakeForm.tsx:158]: Not sure that this is the best place to call + assign data, might be worth pulling it out into a provider. Maybe limit calls too using the `FetchState` format that we sometimes use

File: libs/base/src/lib/components/wallet/PendingTransactions.tsx [hunk 2]
```diff
@@ -89,18 +93,81 @@ const NativeTokenBalance: FC = () => {
...
+  const [isStakeSignedForm, setIsStakeSignedForm] = useState<boolean>(false)
+  const stakeSignedFunctions = ['compoundRewards', 'stake(uint256)']
+
+  useEffect(() => {
+    const fetchSignature = async () => {
+      const address = await signer?.getAddress()
+      if (!address) return
+
+      const walletAddress = await signer?.getAddress()
+      if (!walletAddress) return
```
Comment (chrisjgf) [inline libs/base/src/lib/components/wallet/PendingTransactions.tsx:151]: Duplicated logic?

File: libs/base/src/lib/components/wallet/PendingTransactions.tsx [hunk 3]
```diff
@@ -89,18 +93,81 @@ const NativeTokenBalance: FC = () => {
...
+      setIsStakeSigned(stakeSignatures[walletAddress] ? true : false)
```
Comment (chrisjgf) [inline libs/base/src/lib/components/wallet/PendingTransactions.tsx:152]:
```suggestion
      setIsStakeSigned(!!stakeSignatures[walletAddress])
```

File: libs/base/src/lib/components/wallet/PendingTransactions.tsx (hunk 2 above)
Comment (rafaelugolini) [reply libs/base/src/lib/components/wallet/PendingTransactions.tsx:151]: ops! 🤦

File: apps/governance/src/app/pages/Stake/StakeForm.tsx (hunk 1 above)
Comment (rafaelugolini) [reply apps/governance/src/app/pages/Stake/StakeForm.tsx:158]: @chrisjgf do you think I should inject this in the [AccountProvider](https://github.com/mstable/mStable-apps/blob/962ec44be263620b60c47beba3aaf86d632ee802/libs/base/src/lib/context/AccountProvider.tsx)?

Comment (chrisjgf) [reply apps/governance/src/app/pages/Stake/StakeForm.tsx:158]: Think that sounds good 👍 Maybe see what @JamesLefrere suggests

Comment (rafaelugolini) [reply apps/governance/src/app/pages/Stake/StakeForm.tsx:158]: perfect, I will check with him

File: apps/governance/src/app/pages/Stake/StakeForm.tsx [hunk 4]
```diff
@@ -138,6 +140,22 @@ export const StakeForm: FC<Props> = ({ className, isMigrating = false }) => {
...
+      const signature = await fetch(`https://api.mstable.org/signature/${address}`)
+
+      const body = await signature.json()
+      setStakeSignatures(prevSignatures => ({
+        ...prevSignatures,
+        [address]: body.signature,
+      }))
+    }
+    fetchSignature()
```
Comment (JamesLefrere) [inline apps/governance/src/app/pages/Stake/StakeForm.tsx:156]: Uncaught promise; fetch can fail, .json can fail

File: apps/governance/src/app/pages/Stake/StakeForm.tsx (hunk 1 above)
Comment (JamesLefrere) [reply apps/governance/src/app/pages/Stake/StakeForm.tsx:158]: I think that's a good idea 👍

File: libs/base/src/lib/components/wallet/PendingTransactions.tsx [hunk 5]
```diff
@@ -89,18 +93,78 @@ const NativeTokenBalance: FC = () => {
...
+  const [isStakeSignedForm, setIsStakeSignedForm] = useState<boolean>(false)
+  const stakeSignedFunctions = ['compoundRewards', 'stake(uint256)']
+
+  useEffect(() => {
+    const fetchSignature = async () => {
+      const walletAddress = await signer?.getAddress()
+      if (!walletAddress) return
+      setIsStakeSigned(!!stakeSignatures[walletAddress])
+    }
+    fetchSignature()
```
Comment (JamesLefrere) [inline libs/base/src/lib/components/wallet/PendingTransactions.tsx:151]: The uncaught promisooooor

File: libs/base/src/lib/components/wallet/StakeValidation.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import React, { FC, useState } from 'react'
+import styled from 'styled-components'
+
+const StakeValidationSC = styled.div`
```
Comment (JamesLefrere) [inline libs/base/src/lib/components/wallet/StakeValidation.tsx:4]: We don't have this naming convention anywhere else, but we can start adding it if you think it's helpful. I just hover to see types but YMMV

File: libs/base/src/lib/components/wallet/StakeValidation.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+interface StakeValidationProps {
+  isStakeSigned: boolean
+  setIsStakeSigned: React.Dispatch<React.SetStateAction<boolean>>
+}
+
+export const SIGN_MSG = `I acknowledge that by staking my MTA, I become a member of the mStable DAO. As a member of the mStableDAO, I am committed to:
+
+- supporting decisions and acting in ways that help the protocol achieve its vision of becoming a piece of decentralised, public
+- financial infrastructure for the world.
```
Comment (JamesLefrere) [inline libs/base/src/lib/components/wallet/StakeValidation.tsx:36]: Seems like a line break we don't want

**Review by JamesLefrere — COMMENTED** [review]
A few comments. Let's move that into the AccountProvider 👍

File: libs/base/src/lib/components/wallet/StakeValidation.tsx (hunk 6 above)
Comment (rafaelugolini) [reply libs/base/src/lib/components/wallet/StakeValidation.tsx:4]: I was using this convention in my last company, and I think it was just a force of habit. I will change it. 
We can revisit to use this convention or not separately.

File: libs/base/src/lib/context/AccountProvider.tsx [hunk 8]
```diff
@@ -249,8 +256,43 @@ const OnboardProvider: FC<{
...
+  useEffect(() => {
+    const fetchSignature = async () => {
+      if (!address) return
+
+      fetch(`${API_ENDPOINT}/signature/${address}`)
+        .then(resp => resp.json())
+        .then(json => {
+          if (json.error) return
+          setStakeSignatures(prevSignatures => {
+            // TODO: I'm getting a weird race condition here with the library, this fix the issue
```
Comment (rafaelugolini) [inline libs/base/src/lib/context/AccountProvider.tsx:280]: @JamesLefrere @chrisjgf any better idea on how to deal with this? :)

File: libs/base/src/lib/context/AccountProvider.tsx [hunk 9]
```diff
@@ -249,8 +256,43 @@ const OnboardProvider: FC<{
...
+              ...JSON.parse(localStorage.getItem('stakeSignatures') || '{}'),
+              ...prevSignatures,
+            }
+            return {
+              ...prevHack,
+              [address]: json.signature,
+            }
+          })
+        })
+        .catch(console.error)
```
Comment (JamesLefrere) [inline libs/base/src/lib/context/AccountProvider.tsx:291]: Should catch `fetchSignature` instead; the fetch is uncaught

File: libs/base/src/lib/context/AccountProvider.tsx (hunk 8 above)
Comment (chrisjgf) [reply libs/base/src/lib/context/AccountProvider.tsx:280]: Hmm, I don't see why you would need to put it into an object first (re `prevHack`). Console log the obj blobs & compare?

File: libs/base/src/lib/context/AccountProvider.tsx [hunk 10]
```diff
@@ -249,8 +256,43 @@ const OnboardProvider: FC<{
         console.error(error)
       })
     }
+
+    fetch(`${API_ENDPOINT}/signature`)
```
Comment (chrisjgf) [inline libs/base/src/lib/context/AccountProvider.tsx:260]: Does this work without an address passed?

Comment (rafaelugolini) [reply libs/base/src/lib/context/AccountProvider.tsx:260]: @chrisjgf yes... when we don't pass an address, we get the signature msg, I added some changes in the API to support this.

Comment (chrisjgf) [reply libs/base/src/lib/context/AccountProvider.tsx:260]: Nice, just checking! 👍

**Review by JamesLefrere — APPROVED** [review]
lgtm ser

**Pull request merged**